### ISSUE TITLE
When mpv window is snapped, drag-border resize should be fullfilled by moving 2 non-snapped sides of window to fully stay in the screen, not going out of it. 

### ISSUE THREAD

**Issue by AjvarXX** [mpv-player/mpv#6588]
![MPV resize suggestion](https://user-images.githubusercontent.com/9568793/55115968-76f95100-50ee-11e9-8f7f-393911290164.png)
### mpv version and platform
0.29.1, Windows 7

### Reproduction steps
1. Launch mpv with snap ON in mpv.conf.
2. Snap window to let's say lower right corner so that right and bottom sides were magnet to screen edges.
3. Drag to resize either left or top border of mpv.
### Expected behavior
Window should resize to Left+Up as this way it doesn't go out of the screen and Right and Down borders are snapped so should remain there.
### Actual behavior
Snapped borders are no longer snapped and mpv window is getting resized out of the screen.

**Comment by zc62** [comment]
Probably `--snap-window` should be used together with `--no-keepaspect-window`.

**Comment by AjvarXX** [comment]
> 
> 
> Probably `--snap-window` should be used together with `--no-keepaspect-window`.

No, other players manage to do the job while keeping the aspect ratio. It's just they increase vertically and horizontally from other 2 sides. Probably easy rules like "if side V2 is snapped then move V1", "If side G2 is snapped then move G1". I am not a coder though.
![MPV resize suggestion](https://user-images.githubusercontent.com/9568793/55115897-4e715700-50ee-11e9-9c2e-867d903d0797.png)

**Comment by zc62** [comment]
I know what you mean. I can reproduce what you said. `--no-keepaspect-window` is just a workaround for now, if you need to drag left or up.

If you don't mind dragging the upper-left corner of the window, mpv already works correctly this way.

**Comment by AjvarXX** [comment]
Oh, yes, as a workaround it might be even quicker to just resize and then drag window back to snap with new position instead of aiming at the very corner.
I recently moved from another player and it's inconvenient to loose small features which you got used to and already expect those if you know what I mean._ Because aiming takes time in no-border and resize-then-move-back doesn't show preview of how much monitor space it will take at the end.
Thanks for offer though.

**Comment by qmega** [comment]
This is kind of a left-field idea rather than an exact answer to your request, but if what you're concerned about is aiming efficiency, you might try [AltDrag](https://stefansundin.github.io/altdrag/) if you haven't already. It reproduces the behaviour of many Linux window managers where you can move and resize windows by holding down Alt (or another key of your choice) and clicking anywhere in the window instead of having to target the title bar or edges/corners.

If you were using that, I think you could resize the way you want here by just aiming in the general upper-left quadrant of the window and Alt-dragging with the right mouse button, which is probably even easier than the edge-grabbing you were doing before.

**Comment by AjvarXX** [comment]
> 
> 
> ...you might try [AltDrag](https://stefansundin.github.io/altdrag/) if you haven't already.  If you were using that, I think you could resize the way you want here by just aiming in the general upper-left quadrant of the window and Alt-dragging with the right mouse button, which is probably even easier than the edge-grabbing you were doing before.

Thanks, I will use it for general snapping opened windows and putting those On Top from now on.
But for the aforementioned problem of resize, unfortunately AltDrag ignores keepaspect argument from mpv so that using it will bring black bars.

**Comment by qmega** [comment]
> AltDrag ignores keepaspect argument from mpv so that using it will bring black bars.

Hmm, that seems like something that either mpv or AltDrag should be able to fix. I don't have a Windows dev setup, though.

I took a quick look at our resizing-with-aspect-ratio code[1], though, and your original request might actually not be too hard. We're already deciding which border to move based on where the window is being dragged from. Currently that's just a static mapping[2]. But we also keep track of whether we're currently snapped, so it looks like we could use that.

Basically, when dragging an edge -- giving us two choices of the other edge to adjust to maintain aspect ratio -- if `w32->snapped` and some edge of our window's rect is touching the working area's, pick the opposite edge to adjust.

@pavelxdd It looks like the snapping code is yours. Does the above sound reasonable to you?

[1] https://github.com/mpv-player/mpv/blob/fcb320fd3fd3b1fc5549c009f23fc831bd37a495/video/out/w32_common.c#L998-L1019
[2] https://github.com/mpv-player/mpv/blob/fcb320fd3fd3b1fc5549c009f23fc831bd37a495/video/out/w32_common.c#L206-L221

**Comment by pavelxdd** [comment]
I think that `get_resize_border` could check things like `w32->snapped` (no need to check keepaspect or fullscreen states, since it's already done in WM_SIZING) and window position relative to working area (`get_working_area(w32)`). I don't have Windows dev setup to test this, too.

**Comment by AjvarXX** [comment]
Thank you for looking into this. I can't build mpv build for myself but I will happily download and test anything proposed for this.

**Comment by oakkitten** [comment]
also it would be nice if `Alt-0/1/2` (`set window-scale`) behaved similarly, that is, if a window is snapped, changing sizes shouldn't unsnap the window (neither by pushing it off-screen or making it appear like window 1 in the image)

additionally, auto snapping when pressing `Alt-0/1/2` would be nice; when increasing window scale, the window should attempt to stay inside the monitor
![image](https://user-images.githubusercontent.com/1710718/56422804-017a3e00-62b2-11e9-8af7-00d0107169c8.png)

**Comment by qmega** [comment]
This should get the `developer-needed` and `ms-windows` labels. It's a valid feature request and looks pretty easy but needs someone with a Windows dev setup. This would actually be a good first project for someone new to the mpv codebase, I think.
